### PR TITLE
Fixed #32930 -- Fixed URLValidator when port numbers < 10.

### PR DESCRIPTION
**Pull request by whtsky**
URL Standards requires port number to be null or a 16-bit unsigned integer

### CODE REVIEW

File: django/core/validators.py [hunk 1]
```diff
@@ -87,7 +87,7 @@ class URLValidator(RegexValidator):
         r'^(?:[a-z0-9.+-]*)://'  # scheme is validated separately
         r'(?:[^\s:@/]+(?::[^\s:@/]*)?@)?'  # user:pass authentication
         r'(?:' + ipv4_re + '|' + ipv6_re + '|' + host_re + ')'
-        r'(?::\d{2,5})?'  # port
+        r'(?::\d+)?'  # port
```
Comment (felixxm) [inline django/core/validators.py:90]: Using `\d` instead of `\d{1,5}` is risky and would create a security issue :fire: . The previous solutions raises a `ValidationError` almost instantly on crafted URLs, e.g.
```pycon
>>> from django.core.validators import URLValidator
>>> x = 'https://code.djangoproject.com:' + '9' * (2**24)
>>> URLValidator()(x)
...
  File "/django/django/core/validators.py", line 51, in __call__
    raise ValidationError(self.message, code=self.code, params={'value': value})
django.core.exceptions.ValidationError: <exception str() failed>
```
with this patch I waited few minutes and gave up.

File: tests/validators/valid_urls.txt [hunk 2]
```diff
@@ -25,6 +27,10 @@ https://www.example.com/foo/?bar=baz&inga=42&quux
 http://✪df.ws/123
 http://userid@example.com
 http://userid@example.com/
+http://userid@example.com:0000000000000000080
+http://userid@example.com:0000000000000000080/
```
Comment (felixxm) [inline tests/validators/valid_urls.txt:31]: I don't think that we really want to support such ports, leading zeros are always tricky.

**Review by felixxm — COMMENTED** [review]
@whtsky Thanks for this patch :+1:

File: django/core/validators.py (hunk 1 above)
Comment (whtsky) [reply django/core/validators.py:90]: Should I add `'https://code.djangoproject.com:' + '9' * (2**24)` to the test case, so other developers can realize this case?

Comment (whtsky) [reply django/core/validators.py:90]: it's a very, very long URL.(~16MB in size).

**Comment by carltongibson** [comment]
I wonder if this is worth the complexity. 

* Maybe the `\d{1,5}` change to support port numbers `1-9` — **but** is this an actual issue or just a theoretical one? (Leading to...) 
* Any of the other examples look unrealistic. If a user **really** needed those, I'd be inclined to say that they should implement a custom validator, rather than risk opening a _can of worms_. 

I'm just not sure there's a real use-case to back up the change here. 

🤔

**Comment by whtsky** [comment]
> this an actual issue or just a theoretical one?

I wouldn’t believe people uses port 1-9 for http service until I was assigned to a bug ticket: https://qyfkxt.nxyjglt.nx.gov.cn:8/ 

> Any of the other examples look unrealistic

they’re valid url by definition, and you can actually visit them using browser or curl.
I’d still like to support them since we’re implementing URLValidator, not “PrettyURLValidator” or “CommonURLValidator” — the name indicates it should make as less false negatives as possible.

I think I’ll just drop the port range validation since RFC 3986 does not mandates port number to be a 16-bit uint, and port range validating is causing performance issues.

**Comment by felixxm** [comment]
>  I’d still like to support them since we’re implementing URLValidator, not “PrettyURLValidator” or “CommonURLValidator” — the name indicates it should make as less false negatives as possible.

That's actually not true. `URLValidator` like `EmailValidator` should cover most of cases, we don't want to complicate regular expression (and create potential security issue and false positives) just to cover a few extremely rare cases that are used by very limited number of users.

**Comment by carltongibson** [comment]
Also, we already limit to just a few common `schemes`. We're explicitly **not** trying to match every (and just every) possible URL. 

For more obscure cases folks should I implement a custom validator for their application.

**Comment by whtsky** [comment]
I've updated the regex to `0*\d{1,5}`, this should fix the security issue as well as trying to match more correct url.

**Comment by cjerdonek** [comment]
> I've updated the regex to 0*\d{1,5}, this should fix the security issue as well as trying to match more correct url.

The ticket is just about accepting port numbers < 10, so I'm curious why the change about leading zeros is being included. At the least, it seems like it should be in a separate commit, and maybe even a separate ticket / discussion.

**Comment by felixxm** [comment]
> > I've updated the regex to 0*\d{1,5}, this should fix the security issue as well as trying to match more correct url.
> 
> The ticket is just about accepting port numbers < 10, so I'm curious why the change about leading zeros is being included. At the least, it seems like it should be in a separate commit, and maybe even a separate ticket / discussion.

I left https://github.com/django/django/pull/14650#discussion_r673887103 about leading zeros. IMO we shouldn't support them, they can be confusing and I don't really see any issue in avoiding them.

**Comment by whtsky** [comment]
>  I'm curious why the change about leading zeros is being included. At the least, it seems like it should be in a separate commit, and maybe even a separate ticket / discussion.

I'll remove the support for leading zeros. Should port range check be removed too?

File: tests/validators/invalid_urls.txt [hunk 3]
```diff
@@ -2,6 +2,8 @@ foo
 http://
 http://example
 http://example.
+http://example.com:-1
```
Comment (felixxm) [inline tests/validators/invalid_urls.txt:5]: This raises a `ValidationError` without the patch, please move it to a separate commit.

File: tests/validators/tests.py [hunk 4]
```diff
@@ -258,6 +258,9 @@
     # Trailing junk does not take forever to reject
     (URLValidator(), 'http://www.asdasdasdasdsadfm.com.br ', ValidationError),
     (URLValidator(), 'http://www.asdasdasdasdsadfm.com.br z', ValidationError),
+    # Very long port does not take forever to verify
+    (URLValidator(), 'https://code.djangoproject.com:' + '9' * (2**24), ValidationError),
+    (URLValidator(), 'https://code.djangoproject.com:' + '0' * (2**24) + '1', None),
```
Comment (felixxm) [inline tests/validators/tests.py:263]: I don't see much value in these tests unless we can estimate and check the actual time. It's not "forever" like in the above examples, it will take few minutes instead of less than 1-2 seconds. There is no chance anyone would notice it in the test results.

Comment (whtsky) [reply tests/validators/tests.py:263]: I did some simple search and didn't find a proper way to add timeout to unittest without introducing extra dependency or add bunch of new code. Guess I'll remove these test.

**Comment by felixxm** [comment]
> I'll remove the support for leading zeros. Should port range check be removed too?

I would like to discuss it separately as there are some cases that match `\d` but may crash when accessing `urlsplit()`'s `port`, e.g. emoji numbers:8️8️8️8️. I'm not sure we want to support them but ...

**Comment by whtsky** [comment]
> may crash when accessing urlsplit()'s port, e.g. emoji numbers:8️8️8️8️.

I believe emoji numbers are not supported in any browsers, and they're rejected by URLValidator now as `ValueError` will be thrown when accessing `split_result.port`, and we're catching `ValueError` and raising `ValidationError` for this: https://github.com/django/django/pull/14650/files#diff-d9609d8dc8482b30eac30df16213cba134562949fd62c97573927b89e880f85bR153

**Comment by felixxm** [comment]
> I believe emoji numbers are not supported in any browsers, and they're rejected by URLValidator now as `ValueError` will be thrown when accessing `split_result.port`, and we're catching `ValueError` and raising `ValidationError` for this: https://github.com/django/django/pull/14650/files#diff-d9609d8dc8482b30eac30df16213cba134562949fd62c97573927b89e880f85bR153

They are not rejected without this patch, that's what I'm talking about.

**Comment by whtsky** [comment]
I got your point. Removed port range check.

File: django/core/validators.py [hunk 5]
```diff
@@ -129,7 +129,7 @@ def __call__(self, value):
                 raise
         else:
             # Now verify IPv6 in the netloc part
-            host_match = re.search(r'^\[(.+)\](?::\d{2,5})?$', urlsplit(value).netloc)
+            host_match = re.search(r'^\[(.+)\](?::\d{1,5})?$', urlsplit(value).netloc)
```
Comment (felixxm) [inline django/core/validators.py:132]: This change was untested. I added a test case.

**Review by felixxm — COMMENTED** [review]
@whtsky Thanks for updates :+1: I pushed small edits.
